### PR TITLE
ci: enable microbenchmark

### PR DESCRIPTION
**Pull request by v1v**
### What

Enable microbenchmark to be orchestrated in GitHub actions. Though it will run in Buildkite since the current pool of bare-metals is at the moment provided by Buildkite.

Remove the implementation in Jenkins

### Why

Jenkins will be decommissioned in the near future.

### Test 

See https://github.com/elastic/apm-agent-python/actions/runs/4956518071/jobs/8867291350 that was triggered for the commit https://github.com/elastic/apm-agent-python/pull/1829/commits/02cf84cdd63d9720737363da81192a4f74223d41

image

### Important

This will only run for the `main` branch. If required support for PRs that's something to be done by request in a follow up. There are some security concerns when running PRs accessing any Elastic CI environments.

### CODE REVIEW

File: .github/workflows/microbenchmark.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+      - feature/use-gh-actions-only
+    paths-ignore:
+      - '**.md'
+      - '**.asciidoc'
+    tags:
+      - "v*.*.*"
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:13]: unrequired

File: .github/workflows/microbenchmark.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: microbenchmark
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+      - feature/use-gh-actions-only
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:8]: for testing purposes

File: .github/workflows/microbenchmark.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        uses: elastic/apm-pipeline-library/.github/actions/buildkite@current
+        with:
+          vaultUrl: ${{ secrets.VAULT_ADDR }}
+          vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+          vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+          pipeline: apm-agent-microbenchmark
+          waitFor: true
+          printBuildLogs: true
+          buildEnvVars: |
+            script=scripts/ci/bench.sh
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:36]:
```suggestion
            script=.ci/bench.sh
```

File: .ci/bench.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	export BRANCH_NAME
+fi
+
+## Bench specific
+LANG='C.UTF-8'
+LC_ALL="${LANG}"
+export LANG
+export LC_ALL
+
+./scripts/run-benchmarks.sh "$(realpath ~)" "${APM_AGENT_GO_ES_URL}" "${APM_AGENT_GO_ES_USER}" "${APM_AGENT_GO_ES_PASS}"
```
Comment (v1v) [inline .ci/bench.sh:25]:
```suggestion
# APM_AGENT_PYTHON* env variables are provided by the Buildkite hooks.
./scripts/run-benchmarks.sh "$(realpath ~)" "${APM_AGENT_PYTHON_ES_URL}" "${APM_AGENT_PYTHON_ES_USER}" "${APM_AGENT_PYTHON_ES_PASS}"
```

File: .ci/bench.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	# that's normally happening when running in the CI
+	# owned by Elastic.
+	if [ -z "$HOME" ] ; then
+		HOME=$(realpath ~)
+		export HOME
+	fi
+
+	# Consumed in scripts/run-benchmarks.sh
+	BRANCH_NAME=${GITHUB_REF##*/}
+	export BRANCH_NAME
```
Comment (v1v) [inline .ci/bench.sh:16]:
```suggestion
	export BRANCH_NAME

	# required when running the benchmark
	PATH=$PATH:$HOME/.local/bin
	export PATH
```

File: .ci/bench.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -exo pipefail
+
+## Buildkite specific configuration
+if [ "$CI" == "true" ] ; then
+	# If HOME is not set then use the current directory
+	# that's normally happening when running in the CI
+	# owned by Elastic.
+	if [ -z "$HOME" ] ; then
+		HOME=$(realpath ~)
```
Comment (v1v) [inline .ci/bench.sh:10]:
```suggestion
		HOME=$BUILDKITE_BUILD_CHECKOUT_PATH
```

File: .ci/bench.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+set -exo pipefail
+
+## Buildkite specific configuration
+if [ "$CI" == "true" ] ; then
+	# If HOME is not set then use the current directory
```
Comment (v1v) [inline .ci/bench.sh:6]:
```suggestion
	# If HOME is not set then use the Buildkite workspace
```

File: .github/workflows/microbenchmark.yml (hunk 2 above)
Comment (v1v) [reply .github/workflows/microbenchmark.yml:8]:
```suggestion
```

File: .github/workflows/microbenchmark.yml (hunk 1 above)
Comment (v1v) [reply .github/workflows/microbenchmark.yml:13]:
```suggestion
```

File: .github/workflows/microbenchmark.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            sha=${{ github.sha }}
+            GITHUB_REF=${{ env.GITHUB_REF }}
+
+      - if: ${{ failure() }}
+        uses: elastic/apm-pipeline-library/.github/actions/slack-message@current
+        with:
+          url: ${{ secrets.VAULT_ADDR }}
+          roleId: ${{ secrets.VAULT_ROLE_ID }}
+          secretId: ${{ secrets.VAULT_SECRET_ID }}
+          channel: "#apm-agent-go"
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:44]:
```suggestion
          channel: "#apm-agent-python"
```

**Comment by basepi** [comment]
Alright, I fixed up one major issue here: https://github.com/elastic/apm-agent-python-benchmarks/pull/6

We didn't add `wrapt` to that requirements.txt after we unvendored it in this repo.

However, I don't think that will fix your `requests` error. I think that is due to an unclean environment. The [script](https://github.com/elastic/apm-agent-python/blob/main/scripts/run-benchmarks.sh) creates the virtualenv in `$HOME/.local` which is likely not the correct place to avoid side effects. We don't install `requests` so we should not be seeing an error from `requests`. I recommend modifying this line to target a directory in a local directory so it's clean.

https://github.com/elastic/apm-agent-python/blob/1e38ec53115edc70c36c6485259733a8cde02ed9/scripts/run-benchmarks.sh#L17

**Comment by v1v** [comment]
>  virtualenv -p python3 "${HOME}/.local" 

since HOME is not available in the CI runner, it uses the current workspace, see https://github.com/elastic/apm-agent-python/pull/1829/files#diff-6378df415c7b7a21e9acdbbb20884e24b46274dc6fc5ba41cd4d849eb13083c9R6-R12

Let me also double check just in case something else is going on

File: .github/workflows/microbenchmark.yml [hunk 9]
```diff
@@ -0,0 +1,47 @@
+name: microbenchmark
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:7]:
```suggestion
      - main
      - feature/use-gh-actions-only
```

File: .github/workflows/microbenchmark.yml [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+permissions:
+  contents: read
+
+jobs:
+  microbenchmark:
+    runs-on: ubuntu-latest
+    # wait up to 1 hour
+    timeout-minutes: 60
+    steps:
+      - id: buildkite
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:22]:
```suggestion
      - run: env | sort
      - id: buildkite
```

File: .github/workflows/microbenchmark.yml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+          vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+          vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+          pipeline: apm-agent-microbenchmark
+          waitFor: true
+          printBuildLogs: true
+          buildEnvVars: |
+            script=.ci/bench.sh
+            repo=apm-agent-python
+            sha=${{ github.sha }}
+            GITHUB_REF=${{ env.GITHUB_REF }}
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:36]:
```suggestion
            GITHUB_REF_NAME=${{ env.GITHUB_REF_NAME }}
```

File: .ci/bench.sh [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+	# If HOME is not set then use the Buildkite workspace
+	# that's normally happening when running in the CI
+	# owned by Elastic.
+	if [ -z "$HOME" ] ; then
+		HOME=$BUILDKITE_BUILD_CHECKOUT_PATH
+		export HOME
+	fi
+
+	# Consumed in scripts/run-benchmarks.sh
+	BRANCH_NAME=${GITHUB_REF##*/}
```
Comment (v1v) [inline .ci/bench.sh:15]:
```suggestion
	# GITHUB_REF_NAME is the short ref name of the branch or tag that triggered the workflow run
	BRANCH_NAME=${GITHUB_REF_NAME##*/}
```

File: .github/workflows/microbenchmark.yml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+          vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+          vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+          pipeline: apm-agent-microbenchmark
+          waitFor: true
+          printBuildLogs: true
+          buildEnvVars: |
+            script=.ci/bench.sh
+            repo=apm-agent-python
+            sha=${{ github.sha }}
+            GITHUB_REF_NAME=${{ env.GITHUB_REF_NAME }}
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:38]:
```suggestion
            GITHUB_REF_NAME=${{ github.ref_name }}
```

File: .github/workflows/microbenchmark.yml [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+permissions:
+  contents: read
+
+jobs:
+  microbenchmark:
+    runs-on: ubuntu-latest
+    # wait up to 1 hour
+    timeout-minutes: 60
+    steps:
+      - run: env | sort
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:23]:
```suggestion
```

File: .ci/bench.sh [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+	# owned by Elastic.
+	if [ -z "$HOME" ] ; then
+		HOME=$BUILDKITE_BUILD_CHECKOUT_PATH
+		export HOME
+	fi
+
+	# Consumed in scripts/run-benchmarks.sh
+	# GITHUB_REF_NAME is the short ref name of the branch or tag that triggered the workflow run
+	BRANCH_NAME=${GITHUB_REF_NAME##*/}
+	export BRANCH_NAME
```
Comment (v1v) [inline .ci/bench.sh:17]:
```suggestion
```

File: .github/workflows/microbenchmark.yml [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+          vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+          vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+          pipeline: apm-agent-microbenchmark
+          waitFor: true
+          printBuildLogs: true
+          buildEnvVars: |
+            script=.ci/bench.sh
+            repo=apm-agent-python
+            sha=${{ github.sha }}
+            GITHUB_REF_NAME=${{ github.ref_name }}
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:37]:
```suggestion
            BRANCH_NAME=${{ github.ref_name }}
```

File: .github/workflows/microbenchmark.yml [hunk 17]
```diff
@@ -0,0 +1,48 @@
+name: microbenchmark
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+      - feature/use-gh-actions-only
```
Comment (v1v) [inline .github/workflows/microbenchmark.yml:8]:
```suggestion
```

**Review by reakaleek — APPROVED** [review]

**Review by kuisathaverat — APPROVED** [review]

**Comment by v1v** [comment]
@basepi , I'd like to get your input 🙏

**Comment by basepi** [comment]
Apologies for the delay, PTO and all that.

**Pull request merged**